internal/controllers/mic_reconciler.go

```diff
@@ -0,0 +1,332 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	kmmv1beta1 "github.com/kubernetes-sigs/kernel-module-management/api/v1beta1"
+	"github.com/kubernetes-sigs/kernel-module-management/internal/mbsc"
+	"github.com/kubernetes-sigs/kernel-module-management/internal/mic"
+	"github.com/kubernetes-sigs/kernel-module-management/internal/utils"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	MICReconcilerName = "MICReconciler"
+
+	moduleImageLabelKey = "kmm.node.kubernetes.io/module-image-config"
+	imageLabelKey       = "kmm.node.kubernetes.io/module-image"
+
+	pullerContainerName = "puller"
+)
+
+// micReconciler reconciles a MIC (moduleimagesconfig) object
+type micReconciler struct {
+	micReconHelper micReconcilerHelper
+	podHelper      pullPodManager
+}
+
+func NewMICReconciler(client client.Client, micAPI mic.MIC, mbscAPI mbsc.MBSC, scheme *runtime.Scheme) *micReconciler {
+	podHelper := newPullPodManager(client, scheme)
+	micReconHelper := newMICReconcilerHelper(client, podHelper, micAPI, mbscAPI, scheme)
+	return &micReconciler{
+		micReconHelper: micReconHelper,
+		podHelper:      podHelper,
+	}
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *micReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&kmmv1beta1.ModuleImagesConfig{}).
+		Owns(&v1.Pod{}).
+		Owns(&kmmv1beta1.ModuleBuildSignConfig{}).
+		Named(MICReconcilerName).
+		Complete(
+			reconcile.AsReconciler[*kmmv1beta1.ModuleImagesConfig](mgr.GetClient(), r),
+		)
+}
+
+func (r *micReconciler) Reconcile(ctx context.Context, micObj *kmmv1beta1.ModuleImagesConfig) (ctrl.Result, error) {
+	res := ctrl.Result{}
+	if micObj.GetDeletionTimestamp() != nil {
+		// [TODO] delete check image and build/sign pods
+		return res, nil
+	}
+
+	pods, err := r.podHelper.listImagesPullPods(ctx, micObj)
+	if err != nil {
+		return res, fmt.Errorf("failed to get the image pods for mic %s: %v", micObj.Name, err)
+	}
+
+	// update the status of the current images in the Spec based on pull pods
+	err = r.micReconHelper.updateStatusByPullPods(ctx, micObj, pods)
+	if err != nil {
+		return res, fmt.Errorf("failed tp update the status for MIC %s based on pull pods: %v", micObj.Name, err)
+	}
+
+	// update the status of the current images in the Spec based on MBSC status
```

Comment:
I think this comment is redundant. The method's name is self-explanatory.

Reply:
fixed